README.md
Stooge
======

Stooge is a fully EventMachine enabled DSL for interacting with an AMQP broker. Heavily inspired by Minion (but fresher and completely asynchronous).

Setup
-----

Assuming you already have an AMQP broker installed:

    $ gem install stooge

You can configure the address to the broker using the ```AMQP_URL``` environment variable, or programmatically like this:

```ruby
Stooge.amqp_url = 'amqp://johndoe:abc123@localhost/my_vhost'
```

The default if not specified is ```amqp://guest:guest@localhost/```.

Example usage
-------------

To process a job add the following to a file called ```worker.rb``` and run it with ```ruby worker.rb```. Stooge will start an EventMachine loop that waits for AMQP messages and processes matching jobs until you send ```SIG_INT``` or ```SIG_TERM``` to it.

```ruby
require 'stooge'

Stooge.job('example.puts') do |args|
  puts args['message']
end
```

To push a job onto a queue you call ```Stooge.enqueue``` with the name of the work queue and the data you want to process. The data needs to be JSON serializable and can be for example a hash.

```ruby
require 'stooge'

EM.run do
  Stooge.enqueue('example.puts', :message => 'Hello, world!')
end
```

Error handling
--------------

When an error is thrown in a job handler, the job is re-queued to be done later and the Stooge process exits. If you define an error handler, however, the error handler is run and the job is removed from the queue.

```ruby
Stooge.error do |exception, handler, payload, headers|
  puts "got an error! #{e}"
end
```

Logging
-------

Stooge logs to stdout via ```puts```. You can specify a custom logger like this:

```ruby
Stooge.logger do |msg|
  puts msg
end
```

Testing
-------

To test the business logic in your job handler you can use the ```Stooge.run_handler``` helper method:

```ruby
Stooge.run_handler('example.work', { :foo => 'bar' }).should == 42
```

The return value is the return value from executing the job handler block.

Author
------

Stooge was created by Niklas Holmgren (niklas@sutajio.se) with help from Martin Bruse (@zond) and released under the MIT license. Stooge is very much inspired by Minion (created by Orion Henry), Resque (created by Chris Wanstrath) and Stalker (created by Adam Wiggins).


HISTORY.md

### 0.1.4 (2012-04-18)

* Changed Stooge.enqueue to publish persistent messages by default

### 0.1.3 (2012-04-04)

* Run work handlers in a fiber

### 0.1.2 (2012-03-16)

* Made the default logger flush properly

### 0.1.1 (2012-03-16)

* Added a test suite

### 0.1.0 (2012-03-14)

* Implemented same basic feature set as Minion (i.e. work queues)
* Added ability to easily test job handlers
* Made both #enqueue and #job asynchronous
* Changed from using Bunny to the using official AMQP gem
* Improved error handling logic and logging
* Handling of message content types (JSON is still default)
* Added reconnect to broker on connection failure support


spec/stooge_spec.rb
dir = File.dirname(File.expand_path(__FILE__))
$LOAD_PATH.unshift dir + '/../lib'

require 'stooge'
require 'rspec'

describe Stooge do

  before do
    Stooge.amqp_url = 'amqp://guest:guest@localhost/'
    Stooge.clear_handlers
  end

  it 'should use a default URL to the AMQP broker' do
    Stooge.amqp_url.should == 'amqp://guest:guest@localhost/'
    Stooge.send(:amqp_config).should == {
      :vhost => '/',
      :host => 'localhost',
      :user => 'guest',
      :port => 5672,
      :pass => 'guest'
    }
  end

  it 'should be possible to configure the URL to the AMQP broker' do
    Stooge.amqp_url = 'amqp://john:doe@example.com:5000/vhost'
    Stooge.amqp_url.should == 'amqp://john:doe@example.com:5000/vhost'
    Stooge.send(:amqp_config).should == {
      :vhost => '/vhost',
      :host => 'example.com',
      :user => 'john',
      :port => 5000,
      :pass => 'doe'
    }
  end

  it 'can add a job handler (ruby block) for a specified work queue' do
    Stooge.job('test.work') {}
    Stooge.should have_handlers
  end

  it 'can enqueue jobs on a named queue with data and headers if inside an EM synchrony block' do
    EM.synchrony do
      # Here we stub the call to #publish on the default exchange and
      # intercept the published message instead of sending the message to the
      # broker and then testing if the broker receieved a new message, which
      # is very hard to do reliably because of timing issues and because the
      # message is silently dropped by the broker if no queue is bound to the
      # exchange. It also give us the ability to easily test that the message
      # is published with the correct options.
      Stooge.with_channel do |channel|
        channel.default_exchange.should_receive(:publish).with(
          '{"test":"test"}',
          {
            :routing_key => 'test.work',
            :mandatory => true,
            :persistent => true,
            :content_type => 'application/json',
            :headers => { :foo => 'bar' }
          }).and_yield
      end
      Stooge.enqueue('test.work', { :test => 'test' }, :foo => 'bar')
      Stooge.stop!
    end
  end

  it 'can put a job on a queue and that job is the later picked up by a handler and processed' do
    # This test is a bit tricky and might require some explanation. To
    # actually test that this works we need to first create a job handler that
    # sets a global ($work_performed) variable that we can later check to see
    # wether the job has actually been performed. We then open an EM synchrony
    # block in which we start the job handler (i.e. makes it listen to its
    # queue for job messages). We then enqeue a job and later check if the job
    # was performed as expected.
    #
    # To make all this work we need to deal with some nasty timing issues:
    #
    # 1. We need make sure the handler has had enough time to create a queue
    #    and binding before we can publish a message, or the message will
    #    just be silently dropped. This needs to happen in different EM tick.
    #    We therefor add an EM timer to publish the message 1 second later.
    #    The timer requires us to open up yet another synchrony block, because
    #    the timer block does not execute inside the fiber created by
    #    synchrony which it needs to do for Stooge#enqueue to work (because of
    #    the deferable magic happening behind the scenes).
    #
    # 2. When we have published the message we then need to wait a while for
    #    the job handler to pick it up and process it. We need to keep the
    #    EM loop alive for this to happen, so we simply sleep for 1 second
    #    before we shut the EM loop down and check if the job was processed.
    # 
    # NOTE: This all depend on the communication with the broker taking less
    # than 1 second, which should be more than fine in most cases. If this
    # test stops working for some reason the first thing to try should be to
    # increase the wait time and see if it works then.
    #
    $work_performed = false
    Stooge.job('test.work') { $work_performed = true }
    EM.synchrony do
      Stooge.with_channel do |channel|
        Stooge.start_handlers(channel)
      end
      EM.add_timer(1) do
        EM.synchrony do
          Stooge.enqueue('test.work', { :test => 'test' }, :foo => 'bar')
        end
        sleep(1)
        Stooge.stop!
      end
    end
    $work_performed.should == true
  end

  it 'should be possible to specify a custom global error handler' do
    $error = false
    old_handler = Stooge.error_handler
    Stooge.error do |exception, handler, payload, metadata|
      $error = true
    end
    Stooge.job('test.work') { raise }
    Stooge.run_handler('test.work', '')
    Stooge.error(&old_handler)
    $error.should == true
  end

  it 'should be possible to specify a custom logger' do
    $logged = false
    old_logger = Stooge.logger
    Stooge.logger do |msg|
      $logged = true
    end
    Stooge.log('test')
    Stooge.logger(&old_logger)
    $logged.should == true
  end

  it 'should be possible to test a handler' do
    $work_performed = false
    Stooge.job('test.work') do |args, headers|
      $work_performed = true
      args.should == { 'test' => 'test' }
      headers.should == { :foo => 'bar' }
    end
    Stooge.run_handler('test.work', { :test => 'test' }, :foo => 'bar')
    $work_performed.should == true
  end

end


test/stooge_test.rb
require File.expand_path('../../lib/stooge', __FILE__)
require 'test/unit'

#Stooge.logger {}

class StoogeTest < Test::Unit::TestCase
  def setup
    #Stooge.clear!
    $result = -1
    $handled = false
  end

  def with_an_error_handler
    Stooge.error do |e, job_name, args|
      $handled = e.class
      $job_name = job_name
      $job_args = args
    end
  end

  def test_enqueue_and_work_a_job
    EM.synchrony do
      val = rand(999999)
      deferrable = EM::DefaultDeferrable.new
      Stooge.job('test.job') do 
        if val == payload['val']
          $result = val
          deferrable.set_deferred_status :succeeded
        end
      end
      Stooge.enqueue('test.job', :val => val)
      Stooge.check_all
      EM::Synchrony.sync deferrable
      assert_equal val, $result
      AMQP.stop
      EM.stop
    end
  end

  # def test_invoke_error_handler_when_defined
  #   with_an_error_handler
  #   Stooge.job('my.job') { |args| fail }
  #   Stooge.enqueue('my.job', :foo => 123)
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert $handled
  #   assert_equal 'my.job', $job_name
  #   assert_equal({'foo' => 123}, $job_args)
  # end
  # 
  # def test_should_be_compatible_with_legacy_error_handlers
  #   exception = StandardError.new("Oh my, the job has failed!")
  #   Stooge.error { |e| $handled = e }
  #   Stooge.job('my.job') { |args| raise exception }
  #   Stooge.enqueue('my.job', :foo => 123)
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert_equal exception, $handled
  # end
  # 
  # def test_continue_working_when_error_handler_not_defined
  #   Stooge.job('my.job') { fail }
  #   Stooge.enqueue('my.job')
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert_equal false, $handled
  # end
  # 
  # def test_exception_raised_one_second_before_beanstalk_ttr_reached
  #   with_an_error_handler
  #   Stooge.job('my.job') { sleep(3); $handled = "didn't time out" }
  #   Stooge.enqueue('my.job', {}, :ttr => 2)
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert_equal Stooge::JobTimeout, $handled
  # end
  # 
  # def test_before_filter_gets_run_first
  #   Stooge.before { |name| $flag = "i_was_here" }
  #   Stooge.job('my.job') { |args| $handled = ($flag == 'i_was_here') }
  #   Stooge.enqueue('my.job')
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert_equal true, $handled
  # end
  # 
  # def test_before_filter_passes_the_name_of_the_job
  #   Stooge.before { |name| $jobname = name }
  #   Stooge.job('my.job') { true }
  #   Stooge.enqueue('my.job')
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert_equal 'my.job', $jobname
  # end
  # 
  # def test_before_filter_can_pass_an_instance_var
  #   Stooge.before { |name| @foo = "hello" }
  #   Stooge.job('my.job') { |args| $handled = (@foo == "hello") }
  #   Stooge.enqueue('my.job')
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert_equal true, $handled
  # end
  # 
  # def test_before_filter_invokes_error_handler_when_defined
  #   with_an_error_handler
  #   Stooge.before { |name| fail }
  #   Stooge.job('my.job') {  }
  #   Stooge.enqueue('my.job', :foo => 123)
  #   Stooge.prep
  #   Stooge.work_one_job
  #   assert $handled
  #   assert_equal 'my.job', $job_name
  #   assert_equal({'foo' => 123}, $job_args)
  # end

end


lib/stooge/worker.rb
module Stooge
  class Worker

    #
    # Start the Stooge worker that processes jobs.
    #
    def self.run!
      Stooge.log "Starting stooge"

      Signal.trap('INT') { Stooge.stop! }
      Signal.trap('TERM'){ Stooge.stop! }

      Stooge.start!
    end

    #
    # Should the Stooge worker be started?
    #
    # @return [Boolean] true or false
    #
    def self.run?
      Stooge.has_handlers? &&
        File.expand_path($0) == File.expand_path(app_file)
    end

    private

      CALLERS_TO_IGNORE = [
        /\/stooge(\/worker)?\.rb$/,
        /rubygems\/custom_require\.rb$/,
        /bundler(\/runtime)?\.rb/,
        /<internal:/
      ]

      CALLERS_TO_IGNORE.concat(RUBY_IGNORE_CALLERS) if defined?(RUBY_IGNORE_CALLERS)

      def self.caller_files
        cleaned_caller(1).flatten
      end

      def self.caller_locations
        cleaned_caller 2
      end

      def self.cleaned_caller(keep = 3)
        caller(1).
          map    { |line| line.split(/:(?=\d|in )/, 3)[0,keep] }.
          reject { |file, *_| CALLERS_TO_IGNORE.any? { |pattern| file =~ pattern } }
      end

      def self.app_file
        caller_files.first || $0
      end

  end

  #
  # Starte the worker at exit, if appropriate.
  #
  at_exit { Worker.run! if $!.nil? && Worker.run? }
end


lib/stooge/handler.rb
module Stooge
  class Handler

    attr_accessor :queue_name, :queue_options, :block

    #
    # Create a new handler object.
    #
    # @param [String] queue the name of the queue that this handler will pick
    #   jobs from.
    # @param [Hash] options handler options.
    # @option options [Hash] :queue_options Options to use when creating the
    #   queue.
    #
    def initialize(queue, options = {})
      @queue_name = queue
      @queue_options = options[:queue_options] || {}
      @options = options
      @block = lambda {}
    end

    #
    # Start subscribing to the queue that this handler corresponds to. When
    # a message arive; parse it and call the handler block with the data.
    #
    # @param [AMQP::Channel] channel an open AMQP channel
    #
    def start(channel)
      Stooge.log "starting handler for #{@queue_name}"
      channel.queue(@queue_name, @queue_options) do |queue|
        queue.subscribe(:ack => true) do |metadata, payload|
          Stooge.log "recv: #{@queue_name}"
          run(payload, metadata.content_type, metadata.headers)
          metadata.ack
        end
      end
    end

    #
    # Call the handler block. This method will rescue any exceptions the
    # handler block raises and pass them on to the global error handler.
    #
    # @param [String] payload the message payload
    # @param [String] content_type the MIME content type of the message
    #   payload
    # @param [Hash] headers the message headers
    # 
    # @return [Object] the return value of the handler block
    #
    def run(payload, content_type, headers)
      Fiber.new do
        @block.call(
          decode_payload(payload, content_type),
          headers)
      end.resume
    rescue Object => e
      if Stooge.error_handler
        Stooge.error_handler.call(e,self,payload,headers)
      end
    end

    private

      def decode_payload(payload, content_type)
        case content_type
        when 'application/json'
          MultiJson.decode(payload)
        else
          payload.to_s
        end
      end

  end
end


lib/stooge/work_queue.rb
module Stooge
  module WorkQueue

    #
    # Push a job onto a named queue.
    #
    # Example:
    #
    #   Stooge.enqueue('example.work')
    #
    # @param [String] queue_name The name of a work queue or an array of
    #   names that the job will be sent through in sequential order (a
    #   workflow).
    # @param [Object] data The data to send as input to the job. Needs to be
    #   JSON serializable.
    # @param [Hash] headers AMQP headers to include in the job that gets
    #   pushed. Needs to be a hash with key/value pairs.
    #
    def enqueue(queue_name, data, headers = {})
      EM::Synchrony.sync(aenqueue(queue_name, data, headers))
    end

    #
    # Asynchrounous version of enqueue. Yields when the job has been put onto
    # the queue, if a block is given.
    #
    # @param [String] queue_name The name of a work queue or an array of
    #   names that the job will be sent through in sequential order (a
    #   workflow).
    # @param [Object] data The data to send as input to the job. Needs to be
    #   JSON serializable.
    # @param [Hash] headers AMQP headers to include in the job that gets
    #   pushed. Needs to be a hash with key/value pairs.
    #
    # @return [EM::DefaultDeferrable] Returns an EM::DefaultDeferrable object.
    # 
    def aenqueue(queue_name, data, headers = {})
      deferrable = EM::DefaultDeferrable.new
      with_channel do |channel|
        options = {
          :routing_key => queue_name,
          :mandatory => true,
          :persistent => true,
          :content_type => 'application/json',
          :headers => headers
        }
        channel.default_exchange.publish(MultiJson.encode(data), options) do
          Stooge.log("enqueue: #{queue_name}(#{data})")
          yield if block_given?
          deferrable.set_deferred_status :succeeded
        end
      end
      deferrable
    end

    #
    # Creates a job handler for a named queue.
    #
    # Example:
    #
    #   Stooge.job('example.work') do |args,headers|
    #     # Do the work here...
    #   end
    #
    # @param [String] queue The name of the work queue.
    # @param [Proc] blk a {::Proc} that processes the jobs.
    #
    def job(queue, &blk)
      handler = Stooge::Handler.new(queue, :queue_options => { :durable => true })
      handler.block = blk
      add_handler(handler)
    end

  end
end


lib/stooge.rb
require 'uri'
require 'multi_json'
require 'amqp'
require 'em-synchrony'
require 'stooge/version'
require 'stooge/handler'
require 'stooge/work_queue'
require 'stooge/worker'

module Stooge
  extend self
  extend Stooge::WorkQueue

  @@connection = nil
  @@channel = nil
  @@handlers = []
  @@logger = Proc.new do |msg|
    puts "#{Time.now} :stooge: #{msg}"
    STDOUT.flush
  end
  @@error_handler = Proc.new do |exception, handler, payload, headers|
    Stooge.log "#{handler.queue_name} failed: #{exception.inspect}"
    raise exception
  end

  #
  # Will return a URL to the AMQP broker to use. Will get this from the
  # <code>ENV</code> variable <code>AMQP_URL</code> if present, or use the
  # default which is "amqp://guest:guest@localhost/" (the same as the default
  # for a local RabbitMQ install).
  #
  # @return [String] a URL to an AMQP broker.
  #
  def amqp_url
    @@amqp_url ||= ENV["AMQP_URL"] || "amqp://guest:guest@localhost/"
  end

  #
  # Set the URL to the AMQP broker to use. The format of the URL should be
  # "amqp://username:password@hostname/vhost" (vhost and username/password is
  # optional).
  #
  def amqp_url=(url)
    @@amqp_url = url
  end

  #
  # Log message using the Stooge logger.
  #
  # @param [String] msg the message to log.
  #
  def log(msg)
    @@logger.call(msg)
  end

  #
  # Configure a custom logger for Stooge. The block gets yielded with the
  # log message. You can do whatever you want with it. The default behaviour
  # is to simply output to stdout.
  #
  # @param [Proc] block a {::Proc} to yield when a message needs to be logged.
  # @yieldparam [String] msg the message to log.
  #
  def logger(&blk)
    @@logger = blk
  end

  #
  # Configure a global error handler for Stooge jobs. The block gets yielded
  # when a job handler raises an exception. The default error handler simply
  # logs the error and re-raises the exception. You can use this to for
  # example re-queue a failed job or send email notifications when a job
  # fails.
  #
  # If you don't raise an exception in the error handler the job will be
  # acked with the broker and the broker will consider the job done and remove
  # it from the queue. To make sure the job is not lost you can simply
  # re-raise the same exception in your custom handler:
  #
  #   Stooge.error do |exception, handler, payload, headers|
  #     raise exception
  #   end
  #
  # @param [Proc] block a {::Proc} to yield when an error happens.
  # @yieldparam [Exception] exception the exception object raised.
  # @yieldparam [Stooge::Handler] handler the handler that failed.
  # @yieldparam [Object] payload the message payload that was processed when
  #   the handler failed.
  # @yieldparam [Hash] headers the message headers
  #
  def error(&blk)
    @@error_handler = blk
  end

  #
  # The global error handler.
  #
  # @return [Proc] the error handler block.
  # 
  def error_handler
    @@error_handler
  end

  #
  # Start listening to for new jobs on all queues using the specified channel.
  #
  # @param [AMQP::Channel] channel an open AMQP channel
  #
  def start_handlers(channel)
    @@handlers.each { |h| h.start(channel) }
  end

  #
  # Add a new job handler. Used by {Stooge.job}.
  # 
  # @param [Stooge::Handler] handler a handler object
  #
  def add_handler(handler)
    @@handlers << handler
  end

  #
  # Are there any job handlers defined? Used by {Stooge::Worker} to check if
  # it should start a worker process.
  #
  # @return [Boolean] true or false
  #
  def has_handlers?
    @@handlers.empty? == false
  end

  #
  # Remove all job handlers. Mainly useful in tests to create a clean slate
  # for each test case.
  #
  def clear_handlers
    @@handlers = []
  end

  #
  # Execute a handler block without going through AMQP at all. This is a
  # helper method for use in tests. It allows you to test the business logic
  # in a handler block without having to mess with the details of how Stooge
  # works internally.
  #
  # Examples
  #
  #   Stooge.run_handler('example.work', :foo => 'bar').should == 42
  #
  # @param [String] queue_name the name of the handler to run
  # @param [Object] data message data to send to the handler as arguments
  # @param [Hash] headers optional headers to send as second argument to the
  #   handler block
  #
  # @return the return value of the handler block
  #
  def run_handler(queue_name, data, headers = {})
    @@handlers.each do |handler|
      if handler.queue_name == queue_name
        return handler.run(MultiJson.encode(data), 'application/json', headers)
      end
    end
  end

  #
  # Will yield an open and ready connection.
  #
  # @param [Proc] block a {::Proc} to yield an open and ready connection to.
  #
  def with_connection(&block)
    if @@connection.nil? || @@connection.status == :closed
      @@connection = AMQP.connect(amqp_config)
      @@connection.on_tcp_connection_loss do
        Stooge.log "[network failure] trying to reconnect..."
        @@connection.reconnect
      end
      @@connection.on_recovery do
        Stooge.log "connection with broker recovered"
      end
      @@connection.on_error do |ch, connection_close|
         raise connection_close.reply_text
       end
    end
    @@connection.on_open do
      yield @@connection
    end
  end

  #
  # Will yield an open and ready channel.
  #
  # @param [Proc] block a {::Proc} to yield an open and ready channel to.
  #
  def with_channel(&block)
    with_connection do |connection|
      if @@channel.nil? || @@channel.status == :closed
        @@channel = AMQP::Channel.new(connection, AMQP::Channel.next_channel_id, channel_options)
        @@channel.on_error do |ch, channel_close|
          Stooge.log "channel-level exception: code = #{channel_close.reply_code}, message = #{channel_close.reply_text}"
        end
      end
      @@channel.once_open do
        yield @@channel
      end
    end
  end

  # Start a {::Stooge} worker that consumes messages from the AMQP broker.
  def start!
    EM.synchrony do
      with_channel do |channel|
        start_handlers(channel)
      end
    end
  end

  # Will stop and deactivate {::Stooge}.
  def stop!
    EM.next_tick do
      with_channel do |channel|
        channel.close
      end
      @@channel = nil
      with_connection do |connection|
        connection.close
      end
      @@connection = nil
      EM.stop
    end
  end

  private

    # Will return the default options to use when creating channels.
    def channel_options
      @channel_options ||= {
        :prefetch => 1,
        :auto_recovery => true
      }
    end

    # Will return the default options when connecting to the AMQP broker.
    # Uses the URL from {#amqp_url} to construct these options.
    def amqp_config
      uri = URI.parse(amqp_url)
      {
        :vhost => uri.path,
        :host => uri.host,
        :user => uri.user,
        :port => (uri.port || 5672),
        :pass => uri.password
      }
    rescue Object => e
      raise "invalid AMQP_URL: #{uri.inspect} (#{e})"
    end

end
